ir/src/depgraph.rs

```diff
@@ -0,0 +1,433 @@
+    fn from(from: DepGraphSerdeRepr) -> Self {
+        DepGraph {
+            entries: from
+                .entries
+                .iter()
+                .map(|d| {
+                    (
+                        PathBuf::from(&d.path),
+                        if d.has_graph_entry() {
+                            Some(DepGraphEntry {
+                                mtime: FileTime::from_unix_time(d.unix_seconds, d.nanos),
+                                size: d.size,
+                            })
+                        } else {
+                            None
+                        },
+                    )
+                })
+                .collect(),
+        }
+    }
+}
+
+impl From<DepGraph> for DepGraphSerdeRepr {
+    fn from(dg: DepGraph) -> Self {
+        DepGraphSerdeRepr {
+            entries: dg
+                .entries
+                .iter()
+                .map(|e| {
+                    let path = e.0.to_str().expect("Only UTF names please").to_string();
+                    if let Some(dep_entry) = e.1 {
+                        DepGraphEntrySerdeRepr {
+                            path,
+                            unix_seconds: dep_entry.mtime.unix_seconds(),
+                            nanos: dep_entry.mtime.nanoseconds(),
+                            size: dep_entry.size,
+                        }
+                    } else {
+                        DepGraphEntrySerdeRepr {
+                            path,
+                            unix_seconds: 0,
+                            nanos: 0,
+                            size: 0,
+                        }
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+/// SystemTime lacks a platform independent representation we can
+/// depend on so use FileTime's unix_seconds,nanos.
+/// unix_seconds = nanos = size = 0 is used to represent lack of a DepGraphEntry.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct DepGraphEntrySerdeRepr {
+    path: String,
+    unix_seconds: i64,
+    nanos: u32,
+    size: u64,
+}
+
+impl DepGraphEntrySerdeRepr {
+    fn has_graph_entry(&self) -> bool {
+        !(self.unix_seconds == 0 && self.nanos == 0 && self.size == 0)
+    }
+}
+
+pub struct Change {
+    path: PathBuf,
+    updated_entry: DepGraphEntry,
+}
+
+pub enum InitialState {
+    NotChanged,
+    Changed,
+}
+
+impl DepGraphEntry {
+    fn new(path: &Path) -> Result<DepGraphEntry, io::Error> {
+        let metadata = path.metadata()?;
+        Ok(DepGraphEntry {
+            mtime: FileTime::from_system_time(metadata.modified()?),
+            size: metadata.len(),
+        })
+    }
+}
+
+impl DepGraph {
+    pub fn new() -> DepGraph {
+        DepGraph {
+            entries: HashMap::new(),
+        }
+    }
+}
+
+impl Default for DepGraph {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DepGraph {
+    pub fn is_tracked(&self, path: &PathBuf) -> bool {
```

Comment:
I might prefer `contains`; this would be the method name I would look for if I wanted to find this functionality on something that looked vaguely like a collection.

Reply:
Ty!